File: javascript/resistor-color-trio/resistor-color-trio.js
const Bands = {
  black: 0,
  brown: 1,
  red: 2,
  orange: 3,
  yellow: 4,
  green: 5,
  blue: 6,
  violet: 7,
  grey: 8,
  white: 9,
};

export class ResistorColorTrio {
  constructor([colour1, colour2, colour3]) {
    this.checkForValidColours(colour1, colour2, colour3);

    this.value =
      (Bands[colour1] * 10 + Bands[colour2]) * Math.pow(10, Bands[colour3]);

    this.unit = "ohms";
    if (this.value >= 1000) {
      this.unit = "kilo" + this.unit;
      this.value = this.value / 1000;
    }
    this.label = `Resistor value: ${this.value} ${this.unit}`;
  }

  checkForValidColours = (colour1, colour2, colour3) => {
    if (!(colour1 in Bands && colour2 in Bands && colour3 in Bands)) {
      throw Error("invalid color");
    }
  };
}

File: elixir/resistor-color-trio/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @colours [:black, :brown, :red, :orange, :yellow, :green, :blue, :violet, :grey, :white]

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors) do
    tens = Enum.find_index(@colours, fn colour -> colour == Enum.at(colors, 0) end)
    digits = Enum.find_index(@colours, fn colour -> colour == Enum.at(colors, 1) end)
    magnitude = Enum.find_index(@colours, fn colour -> colour == Enum.at(colors, 2) end)

    resistance = (tens * 10 + digits) * Integer.pow(10, magnitude)
    adjust_for_magnitude(resistance)
  end

  defp adjust_for_magnitude(resistance) when resistance > 1_000_000_000,
    do: {resistance / 1_000_000_000, :gigaohms}

  defp adjust_for_magnitude(resistance) when resistance > 1_000_000,
    do: {resistance / 1_000_000, :megaohms}

  defp adjust_for_magnitude(resistance) when resistance > 1000, do: {resistance / 1000, :kiloohms}
  defp adjust_for_magnitude(resistance), do: {resistance, :ohms}
end
